operator-common/src/main/java/io/strimzi/operator/common/StrimziFuture.java

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.common;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.impl.ContextInternal;
+import io.vertx.core.impl.future.FutureInternal;
+import io.vertx.core.impl.future.Listener;
```

Comment:
Would it be possible to isolate the use of Vert.x to the cluster / topic operator? That would allow us to get rid of the Vert.x as dependency in the `operator-common` and thus not pull the dependency into the operators using it which do not use Vert.x.

Reply:
Hmm, I guess optional would be fine, we can control that. If not, maybe we need to reconsider together with @tombentley what does it actually bring. It would for sure still make it possible to use the resource operators without running on Vert.x even if it was still a dependency.

Reply:
It seems that the optional dependency approach isn't going to work without some duplication. I.e. a Java concurrency and a Vert.x version of the resource operators. Would it be fine to keep the transitive dependency on Vert.x, avoiding all direct usage of the APIs other than by the cluster/topic operators? Then, once those two are migrated to use the standard concurrency API, `StrimziFuture` can shed its Vert.x `Future` implementation and the dependency can be made optional, or possibly removed entirely.

Reply:
@tombentley What do you think about this?

Reply:
The option 3) was what I was originally thinking about. But I can see how your approach is much more elegant from the perspective of the CO changes.

One other thing to maybe consider ... I think the User Operator and Topic Operator really manage only 3 resources: KafkaUsers, KafkaTopics and Secrets. I do not think there is anything else. So in theory, one could also simply move the majority of the resource operators to the CO module and address only the Secret operator which is probably the only shared resource among all of them. But I'm not sure how much it really simplifies this since they all inherit from the same parent classes etc.